### PR TITLE
Enable addition and subtraction on two Amounts.

### PR DESCRIPTION
**Pull request by IguanaBen**
I would like to be able to do addition and subtraction of Amounts in the bql language. This code lets python know how to add and subtract Amounts using add/sub functions that already exist but weren't linked to the class. EvalAdd and EvalSub no longer force the output to be Decimal as they can now be Amounts (or maybe Positions as well in the future).  To test, I added a EvalAmount to create an Amount similar to how EvalConstant is defined.

### CODE REVIEW

File: beancount/query/query_compile_test.py [hunk 1]
```diff
@@ -186,6 +186,14 @@ def test_compile_EvalSub(self):
         c_plus = qc.EvalSub(qc.EvalConstant(17), qc.EvalConstant(18))
         self.assertEqual(Decimal, c_plus.dtype)
 
+    def test_compile_EvalAddAmount(self):
+        c_plus = qc.EvalAdd(qc.EvalAmount(Decimal('1.11','USD')), qc.EvalAmount(Decimal('3.55','USD')))
```
Comment (dnicolodi) [inline beancount/query/query_compile_test.py]: This cannot work, the second arguments to `Decimal` are wrong.

File: beancount/query/query_compile_test.py [hunk 2]
```diff
@@ -186,6 +186,14 @@ def test_compile_EvalSub(self):
...
+        self.assertEqual(Decimal, c_plus.dtype)
+
+    def test_compile_EvalSubAmount(self):
+        c_plus = qc.EvalSub(qc.EvalAmount(Decimal('1.11','USD')), qc.EvalAmount(Decimal('3.55','USD')))
```
Comment (dnicolodi) [inline beancount/query/query_compile_test.py]: Here too.

File: beancount/query/query_compile.py [hunk 3]
```diff
@@ -204,13 +215,13 @@ def __init__(self, left, right):
 class EvalAdd(EvalBinaryOp):
 
     def __init__(self, left, right):
-        f = lambda x, y: Decimal(x + y)
+        f = lambda x, y: (x + y)
```
Comment (dnicolodi) [inline beancount/query/query_compile.py]: Redundant parentheses.

File: beancount/query/query_compile.py [hunk 4]
```diff
@@ -204,13 +215,13 @@ def __init__(self, left, right):
...
         super().__init__(f, left, right, Decimal)
 
 class EvalSub(EvalBinaryOp):
 
     def __init__(self, left, right):
-        f = lambda x, y: Decimal(x - y)
+        f = lambda x, y: (x - y)
```
Comment (dnicolodi) [inline beancount/query/query_compile.py]: Here too.

File: beancount/query/query_compile.py [hunk 5]
```diff
@@ -90,6 +91,16 @@ def __init__(self, value):
...
+class EvalAmount(EvalNode):
+    __slots__ = ('decimal','currency')
+
+    def __init__(self, decimal,currency):
+        super().__init__(type(amount.Amount))
+        self.decimal = decimal
+        self.currency = currency
+
+    def __call__(self, _):
+        return amount.Amount(self.decimal,self.currency)
```
Comment (dnicolodi) [inline beancount/query/query_compile.py]: Missing space after `,`. Also, wouldn't it be simpler to have a single `value` class member, construct the `Amount` instance in the `__init__` method and assign it to `self.value` and just return it here?

File: beancount/query/query_compile.py [hunk 6]
```diff
@@ -90,6 +91,16 @@ def __init__(self, value):
     def __call__(self, _):
         return self.value
 
+class EvalAmount(EvalNode):
```
Comment (dnicolodi) [inline beancount/query/query_compile.py]: Is this class used only in the unit tests? If so, why can't you simply use `EvalConstant` and pass an `amount.Amount` instance to its constructor?

File: beancount/core/amount.py [hunk 7]
```diff
@@ -42,6 +42,13 @@ class Amount(_Amount):
     valid_types_number = (Decimal, type, type(None))
     valid_types_currency = (str, type, type(None))
 
+    def __add__(self,other):
+        return add(self,other)
```
Comment (dnicolodi) [inline beancount/core/amount.py]: Missing space after the commas.

File: beancount/core/amount.py [hunk 8]
```diff
@@ -42,6 +42,13 @@ class Amount(_Amount):
...
+
+    def __sub__(self,other):
+        return sub(self,other)
+
+
```
Comment (dnicolodi) [inline beancount/core/amount.py:58]: Extra newline.

File: beancount/query/query_compile_test.py [hunk 9]
```diff
@@ -186,6 +186,14 @@ def test_compile_EvalSub(self):
         c_plus = qc.EvalSub(qc.EvalConstant(17), qc.EvalConstant(18))
         self.assertEqual(Decimal, c_plus.dtype)
 
+    def test_compile_EvalAddAmount(self):
+        c_plus = qc.EvalAdd(qc.EvalAmount(Decimal('1.11','USD')), qc.EvalAmount(Decimal('3.55','USD')))
+        self.assertEqual(Decimal, c_plus.dtype)
```
Comment (dnicolodi) [inline beancount/query/query_compile_test.py]: Uh!? You testing that the result is of type `Deciamal`. Shouldn't it be of type `Amount`?

File: beancount/query/query_compile.py [hunk 10]
```diff
@@ -90,6 +91,16 @@ def __init__(self, value):
...
+    __slots__ = ('decimal','currency')
+
+    def __init__(self, decimal,currency):
+        super().__init__(type(amount.Amount))
```
Comment (dnicolodi) [inline beancount/query/query_compile.py]: `type(amount.Amount)` seems wrong here, it should simply be `amount.Amount`.

**Review by dnicolodi — COMMENTED** [review]
The changes go in the right direction, but there are a few issues. Also, please make sure that the unit test pass.

**Comment by dnicolodi** [comment]
@blais, we need to decide if we want `bean-query` continue to live here or in the dedicated repository.

File: beancount/query/query_compile.py [hunk 11]
```diff
@@ -204,13 +215,13 @@ def __init__(self, left, right):
 class EvalAdd(EvalBinaryOp):
 
     def __init__(self, left, right):
-        f = lambda x, y: Decimal(x + y)
+        f = lambda x, y: (x + y)
         super().__init__(f, left, right, Decimal)
```
Comment (dnicolodi) [inline beancount/query/query_compile.py]: With the above change this could simply be `super().__init__(operator.add, left, right, Decimal)` which can also be slightly faster.

File: beancount/query/query_compile_test.py (hunk 1 above)
Comment (IguanaBen) [reply beancount/query/query_compile_test.py]: Sorry, I ran the unit test but didn't realize I can it on my installed version of beancounter (2,3.4) and not the source I was editing. My latest commit fixes all the errors I introduced in the unit test.

File: beancount/query/query_compile.py (hunk 3 above)
Comment (IguanaBen) [reply beancount/query/query_compile.py]: Personally, I find this easier to read, but happy to change if you like.

File: beancount/query/query_compile.py (hunk 4 above)
Comment (IguanaBen) [reply beancount/query/query_compile.py]: see above

File: beancount/query/query_compile.py (hunk 5 above)
Comment (IguanaBen) [reply beancount/query/query_compile.py]: code is removed.

File: beancount/query/query_compile.py (hunk 6 above)
Comment (IguanaBen) [reply beancount/query/query_compile.py]: I can do and now do. Thanks.

File: beancount/core/amount.py (hunk 8 above)
Comment (IguanaBen) [reply beancount/core/amount.py:58]: fixed

File: beancount/query/query_compile.py (hunk 10 above)
Comment (IguanaBen) [reply beancount/query/query_compile.py]: removed in latest.

File: beancount/query/query_compile.py (hunk 11 above)
Comment (IguanaBen) [reply beancount/query/query_compile.py]: Good point. I made that change in my latest.

File: beancount/query/query_compile.py (hunk 3 above)
Comment (IguanaBen) [reply beancount/query/query_compile.py]: Actually, moot now since I removed the lambda function and replaced with operator.add as you suggested.

**Comment by IguanaBen** [comment]
With these newer commits, multiplication and division with numbers/decimals is also supported, provided the currency matches.
For example:

select 4*value(position)/2.12  + cost(position) + cost(position)/3

File: beancount/core/amount.py [hunk 12]
```diff
@@ -48,6 +48,22 @@ def __add__(self,other):
     def __sub__(self,other):
         return sub(self,other)
 
+    def __mul__(self,other):
+        try:
+            other = Decimal(other)
+        except ValueError:
+            pass
```
Comment (dnicolodi) [inline beancount/core/amount.py]: Why the attempted implicit conversion? The right way to implement this is:
```python
def __mul__(self, other):
    if not isinstance(other, Decimal):
        return NotImplemented
    return mul(self, other)
```
and similarly for `__truediv__`.

Probably `__add__` and `__sub__` should be in the form:
```python
def __add__(self, other):
    if not isinstance(other, Amount):
        return NotImplemented
    return add(self, other)
```

And at this point most of the asserts in the `add()` and `mul()` functions is redundant and maybe open coding them in the method implementations could be worth for performance.

File: beancount/core/amount.py [hunk 13]
```diff
@@ -48,6 +48,22 @@ def __add__(self,other):
...
+        return mul(self,other)
+
+    def __rmul__(self,other):
+        return self.__mul__(other)
```
Comment (dnicolodi) [inline beancount/core/amount.py]: Just `__rmul__ = __mul__` is enough.

File: beancount/query/query_compile.py [hunk 14]
```diff
@@ -204,23 +193,21 @@ class EvalMul(EvalBinaryOp):
...
     def __init__(self, left, right):
         f = lambda x, y: (x * y)
+        resultType = right.dtype
         if left.dtype != right.dtype:
             if left.dtype == amount.Amount:
-               dtype = left.dtype
-            else:
-               dtype = right.dtype
-        super().__init__(f, left, right, dtype)
+               resultType = left.dtype
```
Comment (dnicolodi) [inline beancount/query/query_compile.py]: Multiplying a `Decimal` by an `Amount` is permitted, but the computation of the return type here would get it wrong. Also, we use snake case and not camel case for variables.

File: beancount/query/query_compile_test.py [hunk 15]
```diff
@@ -171,28 +171,28 @@ def test_compile_EvalOr(self):
...
 
     def test_compile_EvalSub(self):
-        c_plus = qc.EvalSub(qc.EvalConstant(17), qc.EvalConstant(18))
+        c_plus = qc.EvalSub(qc.EvalConstant(D(17)), qc.EvalConstant(D(18)))
         self.assertEqual(Decimal, c_plus.dtype)
 
     def test_compile_EvalAddAmount(self):
-        c_plus = qc.EvalAdd(qc.EvalAmount(Decimal('1.11','USD')), qc.EvalAmount(Decimal('3.55','USD')))
-        self.assertEqual(Decimal, c_plus.dtype)
+        c_plus = qc.EvalAdd(qc.EvalConstant(amount.Amount(Decimal('1.11'),'USD')), qc.EvalConstant(amount.Amount(Decimal('3.55'),'USD')))
```
Comment (dnicolodi) [inline beancount/query/query_compile_test.py]: `D('1.11')` instead of `Decimal('1.11')` for consistency with the other tests. Or you could also add a `from beancount.core.amount import A` and simply write `A('1.11 USD')` for brevity.

File: beancount/query/query_compile_test.py [hunk 16]
```diff
@@ -171,28 +171,28 @@ def test_compile_EvalOr(self):
...
+        self.assertEqual(amount.Amount, c_plus.dtype)
 
     def test_compile_EvalSubAmount(self):
-        c_plus = qc.EvalSub(qc.EvalAmount(Decimal('1.11','USD')), qc.EvalAmount(Decimal('3.55','USD')))
-        self.assertEqual(Decimal, c_plus.dtype)
+        c_plus = qc.EvalSub(qc.EvalConstant(amount.Amount(Decimal('1.11'),'USD')), qc.EvalConstant(amount.Amount(Decimal('3.55'),'USD')))
+        self.assertEqual(amount.Amount, c_plus.dtype)
```
Comment (dnicolodi) [inline beancount/query/query_compile_test.py:204]: Tests for multiplication and division are missing.

**Review by dnicolodi — COMMENTED** [review]
Getting closer. Please squash the comments.

File: beancount/core/amount.py (hunk 13 above)
Comment (IguanaBen) [reply beancount/core/amount.py]: Actually, without __rmul__ you can do "Amount * 4" but not "4 * Amount" .

File: beancount/core/amount.py (hunk 12 above)
Comment (IguanaBen) [reply beancount/core/amount.py]: The reason for the converting to Decimal is to support Numbers /  Ints (and I guess floats come from the ride). If I run:

select 4*value(position)

It will error without the conversion because 'Number is not a Decimal'. If you prefer, I could explicitly check for type Number or Decimal (and convert Numbers), but currently the behavior is to try to make the input a Decimal and if that fails it'll pass the original type which will then return a helpful error. If you try to multiply a Position by an Amount, you'll get:

 unsupported operand type(s) for /: 'Position' and 'Amount'

which seems like reasonable behavior. Another option would be to edit the add function itself to support more than just Decimal inputs, but since that function is (potentially) used elsewhere, I thought it best not to touch it. Let me know what you think is best.

**Comment by IguanaBen** [comment]
To clarify what you said about squashing the comments (did you mean commits?). Should I squash all these commits into 1? If so, should I create a new pull request with that commit or force push that on this pull request?

Thanks,

File: beancount/core/amount.py (hunk 7 above)
Comment (IguanaBen) [reply beancount/core/amount.py]: done

File: beancount/query/query_compile.py (hunk 14 above)
Comment (IguanaBen) [reply beancount/query/query_compile.py]: I'm not seeing the issue. If Left=Decimal, and Right=Amount, then the default right.dtype (Amount) will be amount and this will work. If Left=Amount and Right=Decimal, then then left.dtype == amount.Amount will be true and resultType will be left.dtype which is Amount. Am I missing something.

I'll switch to snake_case on the next commit.

File: beancount/core/amount.py (hunk 13 above)
Comment (dnicolodi) [reply beancount/core/amount.py]: What I was trying to say is that you can replace the function definition with the assignment.

File: beancount/core/amount.py (hunk 12 above)
Comment (dnicolodi) [reply beancount/core/amount.py]: AFAIK, the BQL parser only produces `Decimal` instances. If this is not the case, the conversion should be done somewhere in bean query, not here.

**Comment by dnicolodi** [comment]
I meant squash the commits. Force pushing to this PR is enough.

File: beancount/core/amount.py (hunk 12 above)
Comment (dnicolodi) [reply beancount/core/amount.py]: Correction, the BQL parser also emits `int`s. The check should be `isinstance(other, (Decimal, int))` and the implementation should support that: multiplying a `Decimal` by an `int` is supported and well defined.

File: beancount/query/query_compile_test.py (hunk 9 above)
Comment (IguanaBen) [reply beancount/query/query_compile_test.py]: Yeah, this was a mess with EvalAmount converting to an amount, but this is cleaned up now.

File: beancount/core/amount.py (hunk 12 above)
Comment (IguanaBen) [reply beancount/core/amount.py]: Sounds good. Just to clarify, do you think its better in the __add__ function to check the instance and if it is an int, convert to a Decimal? Or check just check and pass Decimal or Int to the add function and then edit the add function to work with type int in addition to Decimal?

Comment (dnicolodi) [reply beancount/core/amount.py]: The fact that the parser emits both `Deciamal` and `int` objects has repercussions on the definition of the return types of the arithmetic operations below and on the definition of the operations. With your patches, `select 1/3` is a legal operation but the reported return type is `int` while it is `float`. It seems that correctly handling arithmetic functions requires some more thinking about data types in BQL.

File: beancount/core/amount.py (hunk 13 above)
Comment (IguanaBen) [reply beancount/core/amount.py]: Ok. Done.

File: beancount/query/query_compile_test.py (hunk 16 above)
Comment (IguanaBen) [reply beancount/query/query_compile_test.py:204]: Added

File: beancount/query/query_compile.py (hunk 14 above)
Comment (dnicolodi) [reply beancount/query/query_compile.py]: In this context `Decimal` and `Amount` are not the only two possible dtypes.

File: beancount/core/amount.py (hunk 12 above)
Comment (IguanaBen) [reply beancount/core/amount.py]: That's a good point. My latest (squashed) commit tries to address this. Instead of handling it in the __mul__ function, I put it in the EvalMul function to check the types and respond accordingly, which I think might be a little cleaner.

File: beancount/query/query_compile_test.py (hunk 15 above)
Comment (IguanaBen) [reply beancount/query/query_compile_test.py]: Done

File: beancount/query/query_compile.py (hunk 14 above)
Comment (IguanaBen) [reply beancount/query/query_compile.py]: I think this is addressed in my latest version.

File: beancount/query/query_compile.py [hunk 17]
```diff
@@ -192,26 +192,41 @@ def __call__(self, context):
 class EvalMul(EvalBinaryOp):
 
     def __init__(self, left, right):
-        f = lambda x, y: Decimal(x * y)
-        super().__init__(f, left, right, Decimal)
+        if (left.dtype == amount.Amount) and (right.dtype in [int,float, Decimal]):
```
Comment (dnicolodi) [inline beancount/query/query_compile.py]: Where is `float` coming from?

File: beancount/query/query_compile.py [hunk 18]
```diff
@@ -192,26 +192,41 @@ def __call__(self, context):
...
+            result_type = amount.Amount
+            super().__init__(operator.mul, left, right, result_type)
+        elif right.dtype == amount.Amount and (left.dtype in [int,float, Decimal]):
+            result_type = amount.Amount
+            super().__init__(operator.mul, left, right, result_type)
+        elif (left.dtype in [int,Decimal]) and (right.dtype in [int,Decimal]):
+            f = lambda x, y: Decimal(x * y)
```
Comment (dnicolodi) [inline beancount/query/query_compile.py]: Why do you force the multiplication between `int` to a `Decimal`?

File: beancount/query/query_compile.py [hunk 19]
```diff
@@ -192,26 +192,41 @@ def __call__(self, context):
...
 
     def __init__(self, left, right):
-        f = lambda x, y: Decimal(x / y)
-        super().__init__(f, left, right, Decimal)
+        if (left.dtype == amount.Amount) and (right.dtype in [int,float, Decimal]):
+            result_type = amount.Amount
+            super().__init__(operator.truediv, left, right, result_type)
+        elif right.dtype in [int,float,Decimal] and (left.dtype in [int,float, Decimal]):
+            result_type = amount.Decimal
+            f = lambda x, y: Decimal(x / y)
```
Comment (dnicolodi) [inline beancount/query/query_compile.py]: This looses precision. I think the correct way to write it is `Decimal(x) / Decimal(y)`

File: beancount/core/amount.py [hunk 20]
```diff
@@ -42,6 +42,28 @@ class Amount(_Amount):
...
+        return mul(self, other)
+
+    __rmul__ = __mul__
+
+    def __truediv__(self, other):
+        try:
+            other = Decimal(other)
+        except ValueError:
+            pass
+        return div(self, other)
```
Comment (dnicolodi) [inline beancount/core/amount.py:57]: The definition of `__mul__` and `__truediv__` are still wrong: no implicit conversions, please. If we add these methods, they need to respect the Python expected semantics.

**Review by dnicolodi — COMMENTED** [review]
Some remaining comments. Also, please make sure to follow the code formatting conventions: there are many missing spaces after commas and redundant parentheses. `pylint` should warn about those. Also the commit message need to be fixed, something like:
```
bean-query: Allow arithmetic operations on data.Amount dtype

This requires to extend the data.Amount class to support arithmetic dunder methods.
```
would do.

File: beancount/query/query_compile.py (hunk 17 above)
Comment (IguanaBen) [reply beancount/query/query_compile.py]: I was thinking the output of 1/3 would be a float just like in python. But since we explicitly converting int/int to a decimal, then this isn't necessary.

File: beancount/query/query_compile.py (hunk 18 above)
Comment (IguanaBen) [reply beancount/query/query_compile.py]: I can break out the int*int case.

File: beancount/query/query_compile.py (hunk 19 above)
Comment (IguanaBen) [reply beancount/query/query_compile.py]: I don't think you lose precision. In the case of Int/Int, the python return will be a float. If you doe 1/3, then if you take the resulting float and multiply it by 3, you get a float with value 1. And if you then convert it to a Decimal, you'll get a decimal of value 1. But if you convert to Decimal before the division, you get: 0.999999:

```python
>>>Decimal(3*(1/3))
Decimal('1')

>>>Decimal(3*(1/3))
Decimal('1')

```

Is there a use-case I'm not thinking of? In the case of int or decimal divided by decimal, I don't think the conversion makes a difference.

File: beancount/core/amount.py (hunk 20 above)
Comment (IguanaBen) [reply beancount/core/amount.py:57]: Ok, I removed the explicit conversions and edited the underlying add/ div functions to allow ints as data types.

**Comment by IguanaBen** [comment]
Ok, more changes. I think I fixed the missing spaces for formatting, but I ran pylint on the files and it didn't report any issues with the spacing before, so its hard to know for sure. Is there a flag for pylint to get it to issue a warning on [a,b,c]?

File: beancount/query/query_compile.py (hunk 19 above)
Comment (dnicolodi) [reply beancount/query/query_compile.py]:
```python
>>> from decimal import Decimal as D
>>> D(1) / D(3)
Decimal('0.3333333333333333333333333333')
>>> D(1 / 3)
Decimal('0.333333333333333314829616256247390992939472198486328125')
```
The first is correct up to the precision of the decimal representation, the second is not.

File: beancount/query/query_compile.py [hunk 21]
```diff
@@ -192,26 +193,45 @@ def __call__(self, context):
 class EvalMul(EvalBinaryOp):
 
     def __init__(self, left, right):
-        f = lambda x, y: Decimal(x * y)
-        super().__init__(f, left, right, Decimal)
+        if (left.dtype == Amount) and (right.dtype in [int, Decimal]):
```
Comment (dnicolodi) [inline beancount/query/query_compile.py]: You don't need the parentheses.

File: beancount/query/query_compile.py [hunk 22]
```diff
@@ -192,26 +193,45 @@ def __call__(self, context):
...
+            result_type = Amount
+            super().__init__(operator.mul, left, right, result_type)
+        elif right.dtype == Amount and (left.dtype in [int, Decimal]):
```
Comment (dnicolodi) [inline beancount/query/query_compile.py]: ...and indeed you don't have them on the left hand side of the `and` here. But you switch the order of the checks. It is more readable if the check for `left.dtype` always comes first.

File: beancount/query/query_compile.py [hunk 23]
```diff
@@ -192,26 +193,45 @@ def __call__(self, context):
...
+            result_type = Amount
+            super().__init__(operator.mul, left, right, result_type)
+        elif (left.dtype == int) and (right.dtype == int):
+            result_type = int
+            super().__init__(operator.mul, left, right, result_type)
+        elif (left.dtype in [int,Decimal]) and (right.dtype in [int,Decimal]):
```
Comment (dnicolodi) [inline beancount/query/query_compile.py]: Space after comma.

File: beancount/query/query_compile.py [hunk 24]
```diff
@@ -192,26 +193,45 @@ def __call__(self, context):
...
+            result_type = Decimal
+            super().__init__(operator.mul, left, right, result_type)
+        else:
+            raise NotImplementedError("Multiply of {} with {} not supported".format(
```
Comment (dnicolodi) [inline beancount/query/query_compile.py]: Other parts of bean-query use `TypeError` for this kind of errors, I would use the same here)

File: beancount/query/query_compile.py [hunk 25]
```diff
@@ -192,26 +193,45 @@ def __call__(self, context):
 class EvalMul(EvalBinaryOp):
 
     def __init__(self, left, right):
-        f = lambda x, y: Decimal(x * y)
-        super().__init__(f, left, right, Decimal)
+        if (left.dtype == Amount) and (right.dtype in [int, Decimal]):
+            result_type = Amount
```
Comment (dnicolodi) [inline beancount/query/query_compile.py]: I am not sure that having a variable assigned to use its value only at the very next line is an improvement in readability. If you feel that the code is hard to follow like this, you can transform it into a lookup table:
```python
dtypes = {
    (Amoun, int): Amount,
    (int, Amount): Amount,
    (Amont, Decimal): Amount,
    (Decimal, Amount): Amount,
    (int, int): int,
    (int, Decimal): Decimal,
    (Decimal, int): Decimal,
    (Decimal, Decimal): Decimal,
}
dtype = dtypes.get((left.dtype, right.dtype))
if dtype is None:
    raise TypeError(...)
super().__init__(operator.mul, left, right, dtype)
```
which has the benefit of being easily checked for correctness and to be easily extensible.

File: beancount/query/query_compile.py (hunk 19 above)
Comment (IguanaBen) [reply beancount/query/query_compile.py]: Interesting. Fixed.

File: beancount/core/amount.py (hunk 12 above)
Comment (IguanaBen) [reply beancount/core/amount.py]: Done.

File: beancount/core/amount.py [hunk 26]
```diff
@@ -161,10 +175,10 @@ def mul(amount, number):
     Returns:
       An Amount, with the same currency, but with 'number' times units.
     """
-    assert isinstance(amount.number, Decimal), (
-        "Amount's number is not a Decimal instance: {}".format(amount.number))
-    assert isinstance(number, Decimal), (
-        "Number is not a Decimal instance: {}".format(number))
+    assert isinstance(amount.number, (int, Decimal)), (
+        "Amount's number is not an Int or Decimal instance: {}".format(amount.number))
```
Comment (dnicolodi) [inline beancount/core/amount.py]: The type name is `int` not `Int` but anyhow, `amount.number` is supposed to be `Decimal`, not `int`.

File: beancount/core/amount.py [hunk 27]
```diff
@@ -161,10 +175,10 @@ def mul(amount, number):
...
+    assert isinstance(number,(int, Decimal)), (
+        "Number is not an Int or Decimal instance: {}".format(number))
```
Comment (dnicolodi) [inline beancount/core/amount.py]: `int` not `Int`

File: beancount/core/amount.py [hunk 28]
```diff
@@ -176,10 +190,10 @@ def div(amount, number):
...
       An Amount, with the same currency, but with amount units divided by 'number'.
     """
-    assert isinstance(amount.number, Decimal), (
-        "Amount's number is not a Decimal instance: {}".format(amount.number))
-    assert isinstance(number, Decimal), (
-        "Number is not a Decimal instance: {}".format(number))
+    assert isinstance(amount.number, (int, Decimal)), (
+        "Amount's number is not an Int or Decimal instance: {}".format(amount.number))
+    assert isinstance(number, (int, Decimal)), (
+        "Number is not an Int or Decimal instance: {}".format(number))
```
Comment (dnicolodi) [inline beancount/core/amount.py]: Same here.

File: beancount/query/query_compile.py (hunk 21 above)
Comment (IguanaBen) [reply beancount/query/query_compile.py]: removed.

File: beancount/query/query_compile.py (hunk 22 above)
Comment (IguanaBen) [reply beancount/query/query_compile.py]: consolidated into one line and reordered so left always first.

File: beancount/query/query_compile.py (hunk 24 above)
Comment (IguanaBen) [reply beancount/query/query_compile.py]: Changed

File: beancount/query/query_compile.py (hunk 25 above)
Comment (IguanaBen) [reply beancount/query/query_compile.py]: That has some benefits.  However it takes 4 conditionals and turns into an array of 8 options, so it also has some drawbacks. For now, I left it with the current structure (but cleaned to up). Let me know if you think this approach is better.

File: beancount/core/amount.py (hunk 26 above)
Comment (IguanaBen) [reply beancount/core/amount.py]: fixed.

File: beancount/core/amount.py (hunk 28 above)
Comment (IguanaBen) [reply beancount/core/amount.py]: fixed

**Comment by dnicolodi** [comment]
If this parch has any future, it is in the beanquery project incarnation of the code. I have some reserves about adding addition and subtraction of Amounts, please see beancount/beanquery#85

**Comment by blais** [comment]
Agreed. Moved to bean-query.  Closing.
